File: p383e/can_construct.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        int charFreq[256] = {0};
        for (char c: magazine) { charFreq[c] ++; }
        for (char c: ransomNote) {
            charFreq[c] --;
            if (charFreq[c] < 0) return false;
        }
        return true;
    }
};
// TEST
struct Test {
    string ransomNote;
    string magazine;
    bool expected;
    void run() {
        Solution sol;
        bool actual = sol.canConstruct(ransomNote, magazine);
        cout << "Ransom note '" << ransomNote 
             << "' can be constructed from magazine '" << magazine
             << "' -> " << boolalpha << actual << endl;
        assert(actual == expected);
    }
};

int main(int argc, char* argv[]) {
    vector<Test> tests = {
        {
            "a", "b", false
        },
        {
            "aa", "ba", false
        },
        {
            "aa", "aba", true 
        },
    };
    for (auto& t: tests) t.run();
    return 0;
}

File: p383e/can_construct.py
from collections import Counter


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        counter = Counter(magazine)
        for c in ransomNote:
            if c not in counter or counter[c] <= 0:
                return False
            counter[c] -= 1
        return True


# TESTS
tests = [("a", "b", False), ("aa", "ab", False), ("aa", "aab", True)]
for t in tests:
    sol = Solution()
    actual = sol.canConstruct(t[0], t[1])
    print(f'canConstruct("{t[0]}", "{t[1]}") ->', actual)
    assert actual == t[2]
